### PR TITLE
feat: add support for Argo Rollouts resource

### PR DESCRIPTION
**Pull request by dnguy078**

### CODE REVIEW

File: config/crd/bases/crd.argoproj.io_rollouts.yaml [hunk 1]
```diff
@@ -0,0 +1,3630 @@
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
```
Comment (dnguy078) [inline config/crd/bases/crd.argoproj.io_rollouts.yaml:2]: I wasn't sure how to add this CRD via kustomize. Wondering if there was a way to autogenerate this into bases folder?

File: internal/api/v1alpha1/cross_version_object_reference.go [hunk 2]
```diff
@@ -22,7 +22,7 @@ type CrossVersionObjectReference struct {
 
 	// Defines the "Kind" of resource being referred to.
 	// +kubebuilder:validation:Required
-	// +kubebuilder:validation:Enum=Deployment;DaemonSet;StatefulSet
+	// +kubebuilder:validation:Enum=Deployment;DaemonSet;StatefulSet;Rollout
```
[the remark this answers is not in the document]
Comment (dnguy078) [reply internal/api/v1alpha1/cross_version_object_reference.go:26]: Updated

**Comment by diranged** [comment]
@dnguy078 Can you add tests?

File: config/crd/bases/crd.argoproj.io_rollouts.yaml (hunk 1 above)
Comment (dnguy078) [reply config/crd/bases/crd.argoproj.io_rollouts.yaml]: I tried to write test to create a Rollout object and then create a PodAccessTemplate pointing to the Rollout and PodAccessRequest. 

```
 failed to get API group resources: unable to retrieve the complete list of server APIs: argoproj.io/v1alpha1: the server could not find the requested resource
```

I read through the docs for this and tried to apply this but it seems that the cluster itself does not know what a Rollout object is when the client attempts to create it.

A workaround was to manually add the crd to the crd base path. 
https://github.com/diranged/oz/blob/main/internal/builders/podaccessbuilder/suite_test.go#L62.
I wasnt sure if theres other options ?

**Comment by diranged** [comment]
@dnguy078 Does https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/using_an_external_type.md help?

**Comment by dnguy078** [comment]
> @dnguy078 Does https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/using_an_external_type.md help?

Not necessarily, this would require using Go's old vendoring folder and referencing the crd yaml file through there.  This implies that crd.yaml is placed in a folder that has an exportable go package. Argo Rollout places their crds into a folder that no go files https://github.com/argoproj/argo-rollouts/tree/master/manifests so we can't simply vendor this in as well.

**Comment by dnguy078** [comment]
I ended up using `go mod download` to be able to grab the crds from rollouts. Unable to use vendor since the crd files were place into a folder without an go files. 

I ended up using a mixture that approach in https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/using_an_external_type.md#prepare-for-testing. Unable to find much documentations beyond this. also [referenced](https://github.com/operator-framework/operator-sdk/issues/4434)

**Review by diranged — COMMENTED** [review]
one comment, then i think it looks good

File: internal/builders/podaccessbuilder/suite_test.go [hunk 3]
```diff
@@ -59,12 +62,21 @@ var _ = BeforeSuite(func() {
 	logf.SetLogger(logger)
 
 	By("bootstrapping test environment")
+
+	var err error
+
+	// grab go mod directory with Argo rollout CRD to be installed into test environment cluster
+	argoCRDPath, err := extractCRDPath("github.com/argoproj/argo-rollouts", "manifests/crds")
```
[the remark this answers is not in the document]
Comment (dnguy078) [reply internal/builders/podaccessbuilder/suite_test.go:69]: this directory is outside of the repo project and its installed as part of the `$GOPATH/pkg`

**Pull request merged**